### ISSUE TITLE
Can't make streams work on a Raspberry Pi 2

### ISSUE THREAD

**Issue by CedricMi** [linvi/tweetinvi#211]
Hi,

When I test the example universal windows app, it works perfectly on my computer, and I can see the streamed tweets, or at least a few of them, in very short succession.

However, the streaming doesn't work with the Rapsberry Pi 2.
I get a System.Net.Exception inTweetinvi.Streams.dll, I get many thread exciting with code 0, and the streamed tweets are not showned.

The first message, Hi xx Welcome on board with Wndows 10 Universal App is shown though.

I suspect the raspberry is overflooded by the tweets and does'nt have time to write on the UI thread or something.

The CPU utilization goes as high as 78 % but then falls quickly. In addition, there is plenty of RAM left. (350 Mb used)

It would be great if you could fix this.
Thanks

**Comment by linvi** [comment]
Hello unfortunately I no longer have a raspberry pi but I will try to buy one. 
In the meantime could you please share your code.

Thanks

**Comment by CedricMi** [comment]
Hi,

I just bought a raspberry pi 3 to see if it makes any difference.
The code I use for testing is yours (example-universal), I just fill the security token for Twitter.

The public stream works well on debug mode on my computer.
It does not print any tweet with the RBP2. I will tell you immediately if its working with the 3, but I fear this is a multithreading issue.
I’ve begun to play with async / await & co but I have only a basic knowledge of this kind of multithreading and UI thread.

I’m preparing a project due next Wednesday / Thursday for a FabLab Festival, so if you have any ideas to test they are much welcome.

The idea would be to show the tweets of a filtered stream on a custom 8x64 led display controlled by the raspberry.

The test code is very similar to yours, but again, seing it didn’t work, I tested directly your example project:
        private void RunFestivalStream()
        {
            var uiDispatcher = Dispatcher;
            var s = Tweetinvi.Stream.CreateFilteredStream();
            s.AddTrack("@FabLabFestival");
            s.AddTrack("@FabLab_Toulouse");
            s.AddTrack("@0x61646d696e");
            s.MatchingTweetReceived += (o, args) =>
            {
#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
                uiDispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
                {
                    tbTweet.Text = args.Tweet.ToString();
                });
#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
            };

            s.StartStreamMatchingAnyConditionAsync();
        }

Thanks for the library, I’m sure I’ll use it again, if not in this project in another.

Cedric


De : linvi [mailto:notifications@github.com]
Envoyé : samedi 30 avril 2016 14:49
À : linvi/tweetinvi <tweetinvi@noreply.github.com>
Cc : Cédric Mialaret <cedric@mialaret.com>; Author <author@noreply.github.com>
Objet : Re: [linvi/tweetinvi] Can't make streams work on a Raspberry Pi 2 (#211)


Hello unfortunately I no longer have a raspberry pi but I will try to buy one.
In the meantime could you please share your code.

Thanks

**Comment by linvi** [comment]
I have also purchased one, hopefully I should receive it tomorrow. 
I know that another project used raspberry pi with streaming so I am just wondering what is going wrong for you.

I will keep you up to date.

**Comment by CedricMi** [comment]
Hi,

So I’ve retested the app with a Raspberry Pi 3 and it still doesn’t work.

I tested with Windows IoT Core 10.0.14.322.1000, which is the latest Windows Insider Preview version with all Windows Updates, with both default controller drivers (Inbox driver and Direct Memory Mapped Driver).
If you Install your Raspberry 3, prepare to spend about 60-90 min with a good network connection waiting for the downloads of the updates (you need to go on the web page of the raspberry and press check for updates), then the reboot for their installation.

The breakpoint I put on: Message.Text = args.Tweet.ToString();
Is never activated, and I see:

The thread 0x76c has exited with code 0 (0x0).
The thread 0x374 has exited with code 0 (0x0).
The thread 0x4e8 has exited with code 0 (0x0).
The thread 0x6b8 has exited with code 0 (0x0).

After the welcome message
i '{0}'. Welcome on board with Windows 10 Universal App!

When I run it on local Machine, the line gets me in the debugger and it works.
When I remove the breakpoint, I still see messages of the time,
The thread 0xNNN has exited with code 0 (0x0).
But much much less, perhaps 2 every 1-2 minutes.


Cheers,
Cedric


        private void RunSampleStream()
        {
            var uiDispatcher = Dispatcher;
            var s = Stream.CreateSampleStream();

            s.TweetReceived += (o, args) =>
            {
#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
                uiDispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
                {
                    Message.Text = args.Tweet.ToString();
                });
#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
            };

            s.StartStreamAsync();
        }


De : linvi [mailto:notifications@github.com]
Envoyé : samedi 30 avril 2016 18:55
À : linvi/tweetinvi <tweetinvi@noreply.github.com>
Cc : Cédric Mialaret <cedric@mialaret.com>; Author <author@noreply.github.com>
Objet : Re: [linvi/tweetinvi] Can't make streams work on a Raspberry Pi 2 (#211)


I have also purchased one, hopefully I should receive it tomorrow.
I know that another project used raspberry pi with streaming so I am just wondering what is going wrong for you.

I will keep you up to date.

**Comment by linvi** [comment]
I have received my raspberry pi 3 but unfortunately I forgot that I needed a usb keyboard which I don't have.

I am ordering one, but in the meantime could you please try to change this line :

``` c#
uiDispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
{
    tbTweet.Text = args.Tweet.ToString();
});

// To Run

uiDispatcher.Run(CoreDispatcherPriority.Normal, () =>
{
    tbTweet.Text = args.Tweet.ToString();
});
```

Otherwise you will create loads of threads in few milliseconds. Maybe Raspberry cannot handle it.

**Comment by CedricMi** [comment]
There is no Run method for Dispatcher.
I'll have more time later.

**Comment by linvi** [comment]
Sorry I did not tried that but here is the idea (tested this time).

``` c#
private void RunSampleStream()
{
    var uiDispatcher = Dispatcher;
    var s = Stream.CreateSampleStream();

    var i = 0;

    s.TweetReceived += async (o, args) =>
    {
        _buffer += $"{args.Tweet.Text}\r\n";

        ++i;

        // Every 10 calls we request the UI to be refreshed.
        if (i == 10)
        {
            i = 0;

            await uiDispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
            {
                Message.Text = args.Tweet.ToString();
            });
        }
    };

    s.StartStreamAsync();
}
```

**Comment by CedricMi** [comment]
I tried something similar but I get the error:
error CS1929: 'IAsyncAction' does not contain a definition for 'GetAwaiter' and the best extension method overload 'AwaitExtensions.GetAwaiter(Task)' requires a receiver of type 'Task'

I should have time to experiment more in a few hours.

**Comment by linvi** [comment]
For this problem you need to add `using System;`.
And then make sure to have `s.TweetReceived += async (o, args) =>`

**Comment by CedricMi** [comment]
Still doesn't print anything for me.
I don't know if this will give you an idea but I discovered the following.

In all versions of RunSampleStream, original and modified, setting a brealpoint on line:
Message.Text = _buffer;

from (modified)
await uiDispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
                    {
                        Message.Text = _buffer;
                    });

or
 Message.Text = args.Tweet.ToString();

from
uiDispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
                {
                    Message.Text = args.Tweet.ToString();
                });

The breakpoint will be reached on local machine.
The breakpoint **is never reached **on the Raspberry.

**Comment by linvi** [comment]
I am not sure what is going wrong for you. I have been trying on Raspberry Pi 3 and everything works perfectly on my side.

Could you try to get the source code of version 0.9.13.0 from github and check if it works.
I have just pushed the changes that we discussed. The version 0.9.13.0 on github is therefore tested on physical Raspberry Pi 3.

**Comment by linvi** [comment]
You can see the code here : https://github.com/linvi/tweetinvi/blob/0.9.13.0/Examplinvi.UniversalApp/MainPage.xaml.cs

**Comment by CedricMi** [comment]
I already gave the Raspberry Pi 3 to the festival's organizers yesterday so I won't have it until next week.

I wanted to test your 0.9.13.0 changes on my Raspberry Pi 2 but now things are even weirder:

it works as it should on x64 - although I had to recompile each and every subproject by hand until I found the right order to build the solution, which is strange.

But with ARM / Remote machine
 var user = User.GetAuthenticatedUser();
returns null, so 
Message.Text = string.Format("Hi '{0}'. Welcome on board with Windows 10 Universal App!", user.Name);  
crashes

No sure what's going on. Since it works in x64 local machine and since I can use the API console tool on Twitter's website, I assume I don't have authorizations problems.

**Comment by linvi** [comment]
I personally compiled in ARM for the Raspeberry Pi. Yesterday I did not have lot of time so I did not try version 0.9.12.2 but I will do tonight.

**Comment by CedricMi** [comment]
Don't hurry for me, as long as the solution I delivered yesterday still works it's fine.
So I rebuilt one by one all the projets for ARM debug and I don't have this null user anymore.

The streaming  still doesn't work though, and now I see in my output 
Exception thrown: 'System.Net.WebException' in Tweetinvi.Streams.dll

Which I don't think I had in 0.9.12

Anyway, this is a test on RBP2, it may well run on RBP3 so don't loose sleep over it.
